File: apps/web/src/lib/api/tasks.ts
import { request } from "./http";

export type TaskKind = "backtest" | "validation" | "ingest" | "copilot";

export interface Task {
  id: string;
  kind: TaskKind;
  title: string;
  status: string;
  progress_step?: string | null;
  created_at: string;
  finished_at?: string | null;
  ref?: string | null;
  strategy_name?: string | null;
  cancelable: boolean;
}

export const tasksApi = {
  // P2.2 统一任务读模型（回测/验证/摄取/研究助手）。
  listTasks: (limit = 50) =>
    request<Task[]>(`/api/v1/tasks?limit=${limit}`),
  getTask: (id: string) => request<Task>(`/api/v1/tasks/${id}`),
  cancelTask: (id: string) =>
    request<Task>(`/api/v1/tasks/${id}/cancel`, { method: "POST" }),
};

File: services/api/routers/tasks.py
"""Unified task center (P2.2): read model + cancel delegation.

Contract: GET /api/v1/tasks, GET /api/v1/tasks/{id}, POST /api/v1/tasks/{id}/cancel.
"""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from fastapi import status as http_status
from sqlalchemy.orm import Session

from services.api.db import get_db
from services.api.schemas import TaskOut
from services.api.services import tasks as tasks_service

router = APIRouter(prefix="/tasks", tags=["tasks"])


@router.get("", response_model=list[TaskOut])
def list_tasks(
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db),
) -> list[TaskOut]:
    """Newest-first unified task list (backtests, validation, ingest, copilot)."""
    return tasks_service.list_tasks(db, limit=limit)


@router.get("/{task_id}", response_model=TaskOut)
def get_task(task_id: UUID, db: Session = Depends(get_db)) -> TaskOut:
    task = tasks_service.get_task(db, task_id)
    if task is None:
        raise HTTPException(status_code=http_status.HTTP_404_NOT_FOUND, detail="任务不存在。")
    return task


@router.post("/{task_id}/cancel", response_model=TaskOut)
def cancel_task(task_id: UUID, db: Session = Depends(get_db)) -> TaskOut:
    return tasks_service.cancel_task(db, task_id)
